nixos/modules/services/networking/whoogle-search.nix

```diff
@@ -0,0 +1,124 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.whoogle-search;
+  whoogle-search = pkgs.whoogle-search;
+  stateDir = "/var/lib/whoogle-search";
+in
+{
+  options = {
+    services.whoogle-search = {
+
+      enable = mkEnableOption "Whether to enable whoogle-search service.";
+
+      user = mkOption {
+        type = types.str;
+        default = "whoogle-search";
+        description = "The user as which to run whoogle-search daemon.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "whoogle-search";
+        description = "The group as which to run whoogle-search daemon.";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 5000;
+        description = ''
+          Port to listen on.
+        '';
+      };
+
+      listenAddress = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Address to listen on for the web interface.
+        '';
+      };
+
+      dataDir = mkOption {
+        type = types.path;
+        default = stateDir;
+        description = ''
+          The directory where whoogle-search will create session files.
+        '';
+      };
+
+      httpsOnly = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enforce HTTPS redirects for all requests.
+        '';
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users.users = mkIf (cfg.user == "whoogle-search") {
+      whoogle-search = {
+        group = cfg.group;
+        uid = config.ids.uids.whoogle-search;
+        home = cfg.dataDir;
+        description = "Whoogle Search daemon user";
+      };
+    };
+
+    users.groups = mkIf (cfg.group == "whoogle-search") {
+      whoogle-search = {
+        gid = config.ids.gids.whoogle-search;
+      };
+    };
+
+
+    systemd.services.whoogle-search = {
+      description = "Whoogle Search";
+      wantedBy = [ "multi-user.target" ];
+      path = [ whoogle-search ];
+
+      preStart = ''
+        mkdir -m 0750 -p ${cfg.dataDir}
+        cp -r ${whoogle-search}/${pkgs.python3.sitePackages}/app/static_runtime ${cfg.dataDir}/static
+        chown -R ${cfg.user}:${cfg.group} ${cfg.dataDir}
+      '';
```

Comment:
The `mkdir` and `chown` commands here can easily fail. The `preStart` script isn't run as `root` so it is easy to pass values which cause this to fail. You're better off only creating the directory if the user leaves the default value, which your logic already handles. Just drop the `mkdir` and `chown` here and you're good to go.

Also... are we absolutely sure the `cp` is required here? Is there no other way to satisfy the software by specifying the path through environment variables, patching, etc...? If not, use `ln -fs` instead of `cp`.

Reply:
I have removed the `mkdir` and `chown`, but it seems I still need to chmod the `${cfg.dataDir}/static/build` directory. The whoogle-search will generate the session static files into this directory during its runtime.

The `${cfg.dataDir}/` must be also writable to create the `session` dir which is also created by whoogle-search at runtime. This is done using the `StateDirectoryMode` set to `0750`.

> Also... are we absolutely sure the `cp` is required here? Is there no other way to satisfy the software by specifying the path through environment variables, patching, etc...? If not, use `ln -fs` instead of `cp`.

I have tried various tweaks to [environment variables](https://github.com/benbusby/whoogle-search/blob/main/app/__init__.py#L42=) (mostly by changing the `STATIC_FOLDER` and `CONFIG_VOLUME`). I also tried patching the behavior of [`BUILD_FOLDER`](https://github.com/benbusby/whoogle-search/blob/main/app/__init__.py#L45=) variable but it seems that's not possible to easily change the code to serve the static files from the directory which is not **under** the `APP_ROOT`. That's why I am doing following symlink hack:

```
$ ls -alh /nix/store/9vg354zwx3m36l72d2r7p6al2k9m5d5g-whoogle-search-0.7.2/lib/python3.9/site-packages/app/static
lrwxrwxrwx 1 root root 30 Jan  1  1970 /nix/store/9vg354zwx3m36l72d2r7p6al2k9m5d5g-whoogle-search-0.7.2/lib/python3.9/site-packages/app/static -> /var/lib/whoogle-search/static
```

Of course, there might be a better solution to this problem.

Thank you for your comments and hints! I really appreciate it!